### PR TITLE
fix inconsistency in TF and CoreML inference outcome

### PR DESCRIPTION
**Pull request by Jarvi-Izana**
This pr includes change:
1. adding a preprocessing (scale factor of 1/255.f) layer for CoreML spec. @jakesabathia2 found this issue.
2. fix extremely slow convergence of TF training by adding data normalization in `dc_data_iterator.cpp`, where each pixel is divided by 255. 
3. fix the issue that CoreML doesn't work consistently with TF since TF is using non-normalized input data for training whereas CoreML normalizes input data before training.
4. put @fareeha 's NN graph conversion from TF to CoreML back.

Issues to be solved in the next step:
1. src/python/turicreate/toolkits/_tf_utils.py is problematic.
2. the TF trains but doesn't update the weights view, as I discovered with @fareeha. This means the initial weights view is the same as the eventual weights view exported to our Cpp backend. That's why we see the prediction probability to be a random number around .5.

__MY SUMMARY__
- [x] model spec test needs to be modified to accommodate change brought from adding a preprocessing layer.
- [x] pass relevant tests of aligning the behavior of CoreML and TF.
- [x] fix inconsistent preprocessing
- [x] fix incorrect weights export
- [x] align pooling padding with TC 5.8
- [x] fix the hardcoded input name in export_coreml
- [x] populate changes to data iterator test
- [x] pass **ALL** relevant DC Cpp test

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -29,7 +29,7 @@ subdirs(
         table_printer
         network
         process
-        vega_renderer
+        # vega_renderer
```
Comment (Jarvi-Izana) [inline test/CMakeLists.txt]: I will clean this up before merge.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 2]
```diff
@@ -156,14 +161,18 @@ std::unique_ptr<model_spec> drawing_classifier::init_model() const {
...
+      if (ii == 2) pad_type = padding_type::VALID;
+
       result->add_pooling(
           /* name                 */ output_name,
           /* input                */ input_name,
           /* kernel_height        */ 2,
           /* kernel_width         */ 2,
           /* stride_height        */ 2,
           /* stride_width         */ 2,
-          /* padding              */ padding_type::VALID,
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: align with TF implemeation

File: src/python/turicreate/test/test_drawing_classifier.py [hunk 3]
```diff
@@ -305,10 +303,19 @@ def test_export_coreml_with_predict(self):
                 prefix = "pretrained" + str(test_number)
             else:
                 prefix = "scratch" + str(test_number)
-            filename = _mkstemp(prefix + ".mlmodel")[1]
+            # filename = _mkstemp(prefix + ".mlmodel")[1]
+            filename = "/Users/guihaoliang/tc_coreml.mlmodel"
```
Comment (TobyRoseman) [inline src/python/turicreate/test/test_drawing_classifier.py]: I think you want to remove this line.

File: src/python/turicreate/test/test_drawing_classifier.py [hunk 4]
```diff
@@ -305,10 +303,19 @@ def test_export_coreml_with_predict(self):
...
             model.export_coreml(filename)
             mlmodel = _coremltools.models.MLModel(filename)
             tc_preds = model.predict(sf)
+            print(tc_preds)
```
Comment (TobyRoseman) [inline src/python/turicreate/test/test_drawing_classifier.py]: Generally you want to remove print statements from unit tests.

File: src/python/turicreate/test/test_drawing_classifier.py [hunk 5]
```diff
@@ -353,14 +361,14 @@ def test_summary(self):
             model.summary()
 
 
-class DrawingClassifierFromScratchTest(DrawingClassifierTest):
-    @classmethod
-    def setUpClass(self):
-        super(DrawingClassifierFromScratchTest, self).setUpClass(
-            warm_start=None)
+# class DrawingClassifierFromScratchTest(DrawingClassifierTest):
```
Comment (TobyRoseman) [inline src/python/turicreate/test/test_drawing_classifier.py]: I don't think you want to comment out this class or the next class.

File: src/python/turicreate/toolkits/_tf_utils.py [hunk 6]
```diff
@@ -258,4 +258,5 @@ def convert_dense_tf_to_coreml(dense_weights):
 
 	"""
 	dense_weights = np.transpose(dense_weights)
-	return np.reshape(dense_weights, (dense_weights.shape[0], dense_weights.shape[1], 1, 1))
+	# , 1, 1))
```
Comment (TobyRoseman) [inline src/python/turicreate/toolkits/_tf_utils.py]: Let's remove commented out code.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 7]
```diff
@@ -107,7 +108,18 @@ def __init__(self, net_params, batch_size, num_classes):
                 self.sess.run(_tf.assign(_tf.get_default_graph().get_tensor_by_name(key+":0"),
                     net_params[key]))
             else:
-                if 'dense' in key:
+                if 'drawing_dense0_weight' in key:
+                    '''
+                    To make output of MXNET pool3 (NCHW) compatible with TF (NHWC).
```
Comment (TobyRoseman) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: I'm confused by this comment. Do you want mean `Core ML` rather than `MXNET`?

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 8]
```diff
@@ -185,17 +197,50 @@ def export_weights(self):
             respective weight of type `numpy.ndarray`.
         """
 
+        from pprint import pprint as pprint
+        pprint("called once")
         net_params = {}
         layer_names = _tf.trainable_variables()
         layer_weights = self.sess.run(layer_names)
+        ff = open("/Users/guihaoliang/tf_dc_weights.txt", 'w') #as f:
```
Comment (TobyRoseman) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: You don't want to check in this line either

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 9]
```diff
@@ -185,17 +197,50 @@ def export_weights(self):
...
                 else:
                     # TODO: Call _utils.convert_conv2d_tf_to_coreml once #2513 is merged.
-                    net_params.update({var.name.replace(":0", ""): _np.transpose(val, (3, 2, 0, 1))})
+                    pprint("{} ".format(var.name), ff)
+                    pprint(val.shape, ff)
+                    pprint(val, ff)
+                    net_params.update(
+                        {var.name.replace(":0", ""): _np.transpose(val, (3, 2, 0, 1))})
+
+        with open("/Users/guihaoliang/tc_tf.txt", 'w') as f:
```
Comment (TobyRoseman) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: You also don't what to check in this line either.

File: src/python/turicreate/test/test_drawing_classifier.py (hunk 4 above)
Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: yeah, I was cleaning the debug scripts

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 7 above)
Comment (Jarvi-Izana) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: I was told that CoreML uses the MXNET format. @fareeha is the original author of this conversion, and I think she has more context than I do.

**Comment by Jarvi-Izana** [comment]
Don't remind me of resolving conflicts. I will resolve it when available. Besides, the file with conflict doesn't prevent you from reviewing the rest.

File: test/CMakeLists.txt (hunk 1 above)
Comment (nickjong) [reply test/CMakeLists.txt]: Please simply do not push local debugging changes to pull requests. It's too error-prone to ask reviewers to ignore certain changes and assume that we'll remember not to include them. Just have these in a separate commit locally that you can stash/cherry-pick/whatever

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 2 above)
Comment (nickjong) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: What does the exported CoreML do here in 5.8? We want to align with that

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 7 above)
Comment (nickjong) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Please don't reference MXNet, which is irrelevant to us now. CoreML does use NCHW

**Comment by nickjong** [comment]
> Don't remind me of resolving conflicts. I will resolve it when available. Besides, the file with conflict doesn't prevent you from reviewing the rest.

I think resolving conflicts makes it easier to review PRs accurately. Please err on the side of reviewer convenience, not your own!

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 10]
```diff
@@ -85,42 +94,57 @@ def __init__(self, net_params, batch_size, num_classes):
...
-                    dense_weights = _utils.convert_dense_coreml_to_tf(net_params[key])
+                if 'drawing_dense0_weight' in key:
+                    '''
+                    To make output of MXNET pool3 (NCHW) compatible with TF (NHWC).
+                    Decompose FC weights to NCHW. Transpose to NHWC. Reshape back to FC.
+                    '''
+                    mxnet_128_576 = net_params[key]
+                    mxnet_128_576 = _np.reshape(mxnet_128_576, (128, 64, 3, 3))
+                    mxnet_128_576 = _np.transpose(mxnet_128_576, (0, 2, 3, 1))
+                    mxnet_128_576 = _np.reshape(mxnet_128_576, (128, 576))
```
Comment (nickjong) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: It looks like this logic was removed in #2514. @shantanuchhabra Do you have context on why? Just want to make sure we're not waffling back and forth between two issues

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp [hunk 11]
```diff
@@ -67,6 +67,9 @@ class data_iterator {
 
     /** Determines results of shuffle operations if enabled. */
     int random_seed = 0;
+
+    // normalization factor for input data
+    float norm_fator = 255.0f;
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.hpp]: `norm_factor`?

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 12]
```diff
@@ -24,12 +24,20 @@
 
 #include <toolkits/drawing_classifier/data_preparation.hpp>
 #include <toolkits/drawing_classifier/drawing_classifier.hpp>
+#include <core/logging/assertions.hpp>
+#include <toolkits/coreml_export/mlmodel_include.hpp>
+#include <core/util/string_util.hpp>
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Let's keep these alphabetized

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 13]
```diff
@@ -24,12 +24,20 @@
...
 
 namespace turi {
 namespace drawing_classifier {
 
 namespace {
 
+using CoreML::Specification::NeuralNetworkLayer;
+using CoreML::Specification::NeuralNetworkPreprocessing;
+using CoreML::Specification::SizeRange;
+using turi::coreml::MLModelWrapper;
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Let's keep these organized with the below. Line 39 here is fully redundant with 41

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp (hunk 11 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: scale (normalization) factor for each pixel value. More acurately, it should be pixel * 1/255.0.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 12 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Sure thing!

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 13 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: yes. I used those for debugging. Thanks for pointing it out. I will do clean up.

**Comment by Jarvi-Izana** [comment]
> > Don't remind me of resolving conflicts. I will resolve it when available. Besides, the file with conflict doesn't prevent you from reviewing the rest.
> 
> I think resolving conflicts makes it easier to review PRs accurately. Please err on the side of reviewer convenience, not your own!

right! I was about head out for some errands before pushing this PR. I just don't want to block people from reviewing the PR.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 12 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: This is the debug relevant headers that I used to print out the protobuf messages. I will trim them.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 14]
```diff
@@ -188,6 +197,11 @@ std::unique_ptr<model_spec> drawing_classifier::init_model() const {
       /* weight_init_fn      */ initializer,
       /* bias_init_fn        */ zero_weight_initializer());
 
+  input_name = std::move(output_name);
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: add one more relu.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 15]
```diff
@@ -156,14 +161,18 @@ std::unique_ptr<model_spec> drawing_classifier::init_model() const {
       ss.str("");
       ss << prefix << "_pool" << ii << _suffix;
       output_name = ss.str();
+
+      auto pad_type =padding_type::SAME;
+      if (ii == 2) pad_type = padding_type::VALID;
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: be aligned with TF NN architecture.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 16]
```diff
@@ -89,28 +97,25 @@ std::unique_ptr<model_spec> drawing_classifier::init_model() const {
...
+  DASSERT_TRUE(features_list.size() == 1);
+
   const std::string prefix{"drawing"};
   // add suffix when needed.
   const std::string _suffix{""};
-  std::string input_name{"features"};
+  std::string input_name{features_list.front().to<flex_string>()};
   std::string output_name;
 
+  result->add_preprocessing(features_list[0].to<flex_string>(), 1 / 255.f);
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: FYI, add a preprocessing layer.

File: test/CMakeLists.txt (hunk 1 above)
Comment (Jarvi-Izana) [reply test/CMakeLists.txt]: That's fair. I will keep this local.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 7 above)
Comment (Jarvi-Izana) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: changed.

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 17]
```diff
@@ -211,6 +211,10 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
     batch_drawings.erase(pend, batch_drawings.end());
   }
 
+  // do noramization on each pixels
+  parallel_for(batch_drawings.begin(), batch_drawings.end(),
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: divide the scale_factor, which normalizes the data.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 18]
```diff
@@ -194,8 +220,23 @@ def export_weights(self):
                 net_params.update({var.name.replace(":0", ""): val})
             else:
                 if 'dense' in var.name:
-                    net_params.update({var.name.replace(":0", ""): _utils.convert_dense_tf_to_coreml(val)})
+                    if 'drawing_dense0_weight' in var.name:
```
Comment (Jarvi-Izana) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: special treatment for dense0.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 10 above)
Comment (Jarvi-Izana) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: This is about the warm start part, which reads the pre-trained weights from our side.

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 19]
```diff
@@ -18,9 +18,9 @@
 CONV_H = 64
 LSTM_H = 200
 DENSE_H = 128
```
Comment (srikris) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Please revert the changes on this file since you haven't actually touched anything.

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 20]
```diff
@@ -18,9 +18,9 @@
...
-
```
Comment (srikris) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Please revert the changes from this file since you didn't intend to touch anything here.

Comment (Jarvi-Izana) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: this line is done by my IDE, which removes the trailing whitespaces (the redundant tab here).

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 2 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: let me check.

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp (hunk 11 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: renamed to scale_factor in order to avoid ambiguity with norm distribution.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 2 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: all paddings of pooling are using 'VALID'

Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: @jakesabathia2 I told you ;-)

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 21]
```diff
@@ -211,6 +214,10 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
     batch_drawings.erase(pend, batch_drawings.end());
   }
 
+  // do noramization on each pixel
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: Fix spelling

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 22]
```diff
@@ -109,8 +110,10 @@ simple_data_iterator::simple_data_iterator(const parameters& params)
 
       // Initialize random number generator.
       random_engine_(params.random_seed)
-
-{}
+{
+  ASSERT_MSG(scale_factor_ > 0,
+             "scale factor of image should be postive float");
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: Check spelling

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 23]
```diff
@@ -211,6 +214,10 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
     batch_drawings.erase(pend, batch_drawings.end());
   }
 
+  // do noramization on each pixel
+  std::for_each(batch_drawings.begin(), batch_drawings.end(),
+                [=](float& x) { x /= scale_factor_; });
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: If we use the term "scale" let's multiply here and initialize to 1.f / 255.f

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 14 above)
Comment (nickjong) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Naming the relu layers independently of the layers they are activating seems weird to me. Could we match the old implementations behavior, which IIRC had layer names like "drawing_dense0" and "drawing_dense0_relu"?

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp (hunk 22 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: good

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp (hunk 21 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: good

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp (hunk 23 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: okay, doable.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 14 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: yes, sounds reasonable.

**Comment by Jarvi-Izana** [comment]
> Let’s make it “ drawing_dense0_relu“ instead of “ drawing_dense0_relu0”. I don’t think the final 0 helps
> 
> Thanks for working through these issues!!

sure! That's an easy change.

File: src/python/turicreate/test/test_drawing_classifier.py [hunk 24]
```diff
@@ -97,7 +97,7 @@ def setUpClass(self, warm_start='auto'):
                 feature=self.feature,
                 max_iterations=1,
                 warm_start=warm_start)
-        self.trains = [self.check_cross_sf, self.stroke_sf]
+        self.trains = [self.check_cross_sf] # , self.stroke_sf]
```
Comment (TobyRoseman) [inline src/python/turicreate/test/test_drawing_classifier.py]: probably want to delete the commented out code here and on the next line.

File: test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx [hunk 25]
```diff
@@ -125,6 +116,15 @@ BOOST_AUTO_TEST_CASE(test_dc_init_model) {
     TS_ASSERT_EQUALS(dense_layer.innerproduct().outputchannels(), 128);
   }
 
+  layer_index++;
+  {
+      const auto& relu_layer = nn.layers(layer_index);
```
Comment (TobyRoseman) [inline test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx]: Does using `auto` make the code more or less readable?

File: test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx [hunk 26]
```diff
@@ -264,6 +267,8 @@ BOOST_AUTO_TEST_CASE(test_save_load) {
   constexpr unsigned int num_classes = 10;
   const std::string target = "target";
   const std::vector<std::string> features = {"0"};
+  // dc enforces the input feature to be 1, in case someday we want more
```
Comment (TobyRoseman) [inline test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx]: `dc` -> `DC' or even better type it all out: `Drawing Classifier`.

File: test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx (hunk 25 above)
Comment (Jarvi-Izana) [reply test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx]: The type is generated by protobuf: 
```
::CoreML::Specification::NeuralNetworkLayer&
```
`auto` can better decouple the underlying protobuf structure from the application layer if we have any change in protobuf message protocol.

Furthermore, the fully qualified name of this type is hard to memorize.

File: src/python/turicreate/test/test_drawing_classifier.py (hunk 24 above)
Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: Yes, once `stroke` based tests are all enabled, we will.

File: test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx (hunk 26 above)
Comment (Jarvi-Izana) [reply test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx]: I will give you that! Thanks for point out.

**Comment by Jarvi-Izana** [comment]
passed all tests locally. Still, some compile error blocks me from using the internal pipeline. I will merge this change anyway and once our build pipeline is unblocked, I will run changes in our internal pipeline to do rest convergence.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 10 above)
Comment (shantanuchhabra) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Looks like removing the code was an oversight on my part. In an effort to reduce code and simplify, I didn't realize that `dense0` was a special case that the conversion utilities in `_tf_utils.py` do not take care of. Apologies!

Comment (nickjong) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Sorry, I missed this during code review too, @fareeha.

When in doubt, please err on the side of understanding/asking about the code you're deleting without replacing, especially when the original author went through the trouble of documenting it as a special case! We lost some development time due to re-engineering this.

File: src/python/turicreate/test/test_drawing_classifier.py [hunk 27]
```diff
@@ -290,8 +290,6 @@ def test_export_coreml(self):
             filename = _mkstemp("bingo.mlmodel")[1]
             model.export_coreml(filename)
 
-    # @unittest.skipIf(_sys.platform != "darwin", "Core ML only supported on Mac")
-    @pytest.mark.xfail(IS_PRE_6_0_RC, reason="Coming soon: Need to test and debug export_to_coreml")
     def test_export_coreml_with_predict(self):
```
Comment (srikris) [inline src/python/turicreate/test/test_drawing_classifier.py]: You commented out the `@unittest.skipIf(_sys.platform != "darwin", "Core ML only supported on Mac")`. You need to add it back. coremltools doesn't support predict on Linux.

Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: good

Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: I did remember that I installed on my linux box before. And now I verified it:
```
coremltools-3.1-cp36-none-manylinux1_x86_64.whl (3.4MB)
```
it supports `manylinux1`.

Comment (srikris) [reply src/python/turicreate/test/test_drawing_classifier.py]: The wheel is supported but the actual predict call in coremltools is only possible on Mac. So the "test_with_predict" only will work on Mac.

Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: I'm afraid this is not the case. It passed all our internal Linux platform pipeline.

Comment (Jarvi-Izana) [reply src/python/turicreate/test/test_drawing_classifier.py]: per offline discussion. It's my mistake of overlooking the build log.

**Comment by Jarvi-Izana** [comment]
Checked internal test pipeline. It looks good to me. Since I do all manual check on the test log, there might be oversight.
